File: clients/android/app/src/main/kotlin/com/lextures/android/core/auth/AuthCallbackParser.kt
package com.lextures.android.core.auth

import java.net.URI
import java.net.URLDecoder

sealed class SsoProvider {
    data class Saml(val idpId: String) : SsoProvider()
    data class Oidc(val path: String, val label: String) : SsoProvider()
}

data class AuthCallbackPayload(
    val accessToken: String? = null,
    val refreshToken: String? = null,
    val expiresIn: Int? = null,
    val requiresMfa: Boolean = false,
    val mfaPendingToken: String? = null,
    val mfaSetupRequired: Boolean = false,
    val magicLinkToken: String? = null,
) {
    fun asTokenResponse(): AuthTokenResponse = AuthTokenResponse(
        accessToken = accessToken,
        refreshToken = refreshToken,
        expiresIn = expiresIn,
        requiresMfa = requiresMfa.takeIf { it },
        mfaPendingToken = mfaPendingToken,
        mfaSetupRequired = mfaSetupRequired.takeIf { it },
        user = null,
    )
}

object AuthCallbackParser {
    fun parse(raw: String?): AuthCallbackPayload? {
        val trimmed = raw?.trim().orEmpty()
        if (trimmed.isEmpty()) return null
        return parseMagicLink(trimmed) ?: parseAuthCallback(trimmed)
    }

    fun parseMagicLink(value: String): AuthCallbackPayload? {
        val path = extractPath(value) ?: return null
        val segments = path.trim('/').split('/').filter { it.isNotEmpty() }
        if (segments.size < 2 || segments[0].lowercase() != "login" || segments[1].lowercase() != "magic-link") {
            return null
        }
        val token = queryValue("token", value)?.trim().orEmpty()
        if (token.isEmpty()) return null
        return AuthCallbackPayload(magicLinkToken = token)
    }

    fun parseAuthCallback(value: String): AuthCallbackPayload? {
        val uri = runCatching { URI(value) }.getOrNull() ?: return null
        val isCallback = uri.scheme.equals(AuthConstants.CALLBACK_SCHEME, ignoreCase = true) &&
            uri.host.equals(AuthConstants.CALLBACK_HOST, ignoreCase = true) &&
            normalizePath(uri.path) == AuthConstants.CALLBACK_PATH
        if (!isCallback) return null
        return payloadFromUri(uri)
    }

    fun payloadFromUri(uri: URI): AuthCallbackPayload {
        val params = mutableMapOf<String, String>()
        parseQuery(uri.rawQuery).forEach { (key, value) -> params[key] = value }
        if (!uri.rawFragment.isNullOrBlank()) {
            parseQuery(uri.rawFragment).forEach { (key, value) -> params[key] = value }
        }
        return AuthCallbackPayload(
            accessToken = params["access_token"],
            refreshToken = params["refresh_token"],
            expiresIn = params["expires_in"]?.toIntOrNull(),
            requiresMfa = params["requires_mfa"] == "1" || params["requires_mfa"] == "true",
            mfaPendingToken = params["mfa_pending_token"],
            mfaSetupRequired = params["mfa_setup_required"] == "1" || params["mfa_setup_required"] == "true",
        )
    }

    private fun normalizePath(path: String?): String {
        if (path.isNullOrBlank()) return ""
        return if (path.startsWith("/")) path else "/$path"
    }

    private fun parseQuery(raw: String?): List<Pair<String, String>> {
        if (raw.isNullOrBlank()) return emptyList()
        return raw.split('&').mapNotNull { pair ->
            val parts = pair.split('=', limit = 2)
            if (parts.isEmpty()) return@mapNotNull null
            val key = parts[0]
            val value = if (parts.size == 2) URLDecoder.decode(parts[1], "UTF-8") else ""
            key to value
        }
    }

    private fun extractPath(value: String): String? {
        if (value.startsWith("lextures://")) {
            val pathPart = value.removePrefix("lextures://").substringBefore('?').substringBefore('#')
            return normalizePath(pathPart)
        }
        if (value.startsWith("/")) return value.substringBefore('?').substringBefore('#')
        if (value.startsWith("http://") || value.startsWith("https://")) {
            val uri = runCatching { URI(value) }.getOrNull() ?: return null
            val host = uri.host?.lowercase().orEmpty()
            if (host == "lextures.com" || host.endsWith(".lextures.com") || host == "localhost") {
                return normalizePath(uri.path)
            }
        }
        return null
    }

    private fun queryValue(name: String, value: String): String? {
        val uri = runCatching { URI(value) }.getOrNull() ?: return null
        return parseQuery(uri.rawQuery).firstOrNull { it.first == name }?.second
    }
}

File: clients/ios/Lextures/Core/Auth/AuthCallbackParser.swift
import Foundation

/// Parsed auth payload from SSO (`lextures://auth/callback`) or magic-link deep links.
struct AuthCallbackPayload: Equatable {
    var accessToken: String?
    var refreshToken: String?
    var expiresIn: Int?
    var requiresMFA: Bool
    var mfaPendingToken: String?
    var mfaSetupRequired: Bool
    var magicLinkToken: String?

    var asTokenResponse: AuthTokenResponse {
        AuthTokenResponse(
            accessToken: accessToken,
            refreshToken: refreshToken,
            expiresIn: expiresIn,
            requiresMFA: requiresMFA ? true : nil,
            mfaPendingToken: mfaPendingToken,
            mfaSetupRequired: mfaSetupRequired ? true : nil,
            user: nil
        )
    }
}

enum AuthCallbackParser {
    /// Returns a payload when `raw` is an auth callback or magic-link URL; otherwise `nil`.
    static func parse(_ raw: String?) -> AuthCallbackPayload? {
        guard let raw, !raw.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return nil }
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)

        if let magic = parseMagicLink(trimmed) {
            return magic
        }
        if let callback = parseAuthCallback(trimmed) {
            return callback
        }
        return nil
    }

    static func parseMagicLink(_ value: String) -> AuthCallbackPayload? {
        guard let path = extractPath(from: value) else { return nil }
        let segments = path.split(separator: "/").map(String.init)
        guard segments.first?.lowercased() == "login",
              segments.count >= 2,
              segments[1].lowercased() == "magic-link" else {
            return nil
        }
        let token = queryValue(named: "token", in: value)?.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let token, !token.isEmpty else { return nil }
        return AuthCallbackPayload(
            accessToken: nil,
            refreshToken: nil,
            expiresIn: nil,
            requiresMFA: false,
            mfaPendingToken: nil,
            mfaSetupRequired: false,
            magicLinkToken: token
        )
    }

    static func parseAuthCallback(_ value: String) -> AuthCallbackPayload? {
        guard let url = URL(string: value) else { return nil }
        let isCustomScheme = url.scheme?.lowercased() == AuthConstants.callbackScheme
            && url.host?.lowercased() == AuthConstants.callbackHost
            && url.path == AuthConstants.callbackPath
        if !isCustomScheme {
            return nil
        }
        return payload(from: url)
    }

    static func payload(from url: URL) -> AuthCallbackPayload {
        var params = queryItems(from: url)
        if let fragment = url.fragment, !fragment.isEmpty {
            params.merge(queryItems(fromFragment: fragment)) { _, new in new }
        }
        return AuthCallbackPayload(
            accessToken: params["access_token"],
            refreshToken: params["refresh_token"],
            expiresIn: params["expires_in"].flatMap { Int($0) },
            requiresMFA: params["requires_mfa"] == "1" || params["requires_mfa"] == "true",
            mfaPendingToken: params["mfa_pending_token"],
            mfaSetupRequired: params["mfa_setup_required"] == "1" || params["mfa_setup_required"] == "true",
            magicLinkToken: nil
        )
    }

    private static func extractPath(from value: String) -> String? {
        if value.hasPrefix("lextures://") {
            let stripped = String(value.dropFirst("lextures://".count))
            return stripped.hasPrefix("/") ? stripped : "/\(stripped)"
        }
        if value.hasPrefix("/") {
            return value.split(separator: "?", maxSplits: 1).first.map(String.init) ?? value
        }
        if let url = URL(string: value), let host = url.host?.lowercased() {
            if host == "lextures.com" || host.hasSuffix(".lextures.com") || host == "localhost" {
                var path = url.path
                if !path.hasPrefix("/") { path = "/\(path)" }
                return path
            }
        }
        return nil
    }

    private static func queryValue(named name: String, in value: String) -> String? {
        guard let url = URL(string: value) else { return nil }
        return queryItems(from: url)[name]
    }

    private static func queryItems(from url: URL) -> [String: String] {
        var out: [String: String] = [:]
        URLComponents(url: url, resolvingAgainstBaseURL: false)?.queryItems?.forEach { item in
            if let value = item.value {
                out[item.name] = value
            }
        }
        return out
    }

    private static func queryItems(fromFragment fragment: String) -> [String: String] {
        var components = URLComponents()
        components.query = fragment
        var out: [String: String] = [:]
        components.queryItems?.forEach { item in
            if let value = item.value {
                out[item.name] = value
            }
        }
        return out
    }
}
